docs/reference/design-patterns.md

```diff
@@ -0,0 +1,372 @@
+Kotlin [`object`](object-declarations.html#object-declarations) has the same semantics:
+
+``` kotlin
+public object Singleton {
+    var data = "any useful shared information"
+}
+
+// Usage:
+val data = Singleton.data
+```
+
+### Delegate (Decorator)
+
+``` java
+interface Printer {
+    void print(String msg);
+}
+
+final class HP3000 implements Printer {
+    void print(String msg) {
+        System.out.println(msg);
+    }
+}
+
+class MyPrinter implements Printer {
+    Printer p = new HP3000();
+
+    void print(String msg) {
+        p.print(msg);
+    }
+}
+
+// Usage:
+MyPrinter mp = new MyPrinter();
+mp.print("Hello!");
+```
+
+[Delegation pattern](https://en.wikipedia.org/wiki/Delegation_pattern) allows inheriting behaviour of a class without extending it.
+This is invaluable when the class, implementing the behaviour, is final.
+Also this technique has no limitation on the number of delegated interfaces, thus providing a mean for [multiple inheritance](https://en.wikipedia.org/wiki/Multiple_inheritance).
+Kotlin cuts the boilerplate here with its [class delegation](https://kotlinlang.org/docs/reference/delegation.html#class-delegation) syntax:
+
+``` kotlin
+// --- interface `Printer` and class `HP3000` are defined in the Java snipet above ---
+
+class MyPrinter : Printer by HP3000()
+
+// Usage:
+val mp = MyPrinter()
+mp.print("Hello!")
+```
+
+The delegate object can be a constructor parameter, allowing runtime behaviour configuration, or acting as *dependency injection*:
+
+``` kotlin
+class ConfigurablePrinter(printer: Printer = HP3000()) : Printer by printer
+
+// Usage:
+val workPrinter = ConfigurablePrinter()
+workPrinter.print("Hello!")
+
+val testPrinter = ConfigurablePrinter(printer = HP555())
+testPrinter.print("test page")
+```
+
+
+## Enhanced patterns
+
+Patterns that may be improved with Kotlin features.
+
+### Builder
+
+``` java
+class Car {
+    final int year;
+    final String model;
+
+    Car(CarBuilder builder) {
+        year = builder.year;
+        model = builder.model;
+    }
+}
+
+class CarBuilder {
+    int year;
+    String model = "default-model";
+
+    void setYear(int year) {
+        this.year = year;
+    }
+
+    void setModel(String model) {
+        this.model = model;
+    }
+}
+
+// Usage:
+CarBuilder cb = new CarBuilder();
+cb.setYear(1908);
+cb.setModel("Model T");
+
+Car car = new Car(cb);
+```
+
+The [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) is used to create objects with `final` fields without providing a constructor with a massive list of arguments.
+Also it allows having default values for `final` fields that are optional to specify.
+All of this is easily achieved in Kotlin with the [primary constructor](classes.html#constructors) syntax.
+Note how using [named arguments](functions.html#named-arguments) makes the syntax well-readable even for longer parameter lists:
+
+``` kotlin
+class Car(val year: Int,
```

Comment:
Hm, I didn't put much thinking into it. The style I use (here and in other projects) does not cause much trouble and IJ auto-formatter supports it. I don't think the docs use one style or another as for now. I believe this is an interesting and very important discussion, and it should be followed by creating of an extensive style guid in the docs.

For now I hope the team merges my patch ASAP, so that others could patch the text easily. I an going to raise the issue after Kotlin 1.0.3 is out